src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -2498,6 +2499,337 @@ public static decimal Average<TSource>(this IEnumerable<TSource> source, Func<TS
+                    while(e.MoveNext())
+                    {
+                        TSource x = e.Current;
+                        TKey xKey = keySelector(x);
+                        if(comparer.Compare(xKey, key) > 0)
+                        {
+                            value = x;
+                            key = xKey;
+                        }
+                    }
+                }
+            }
+            return value;
+        }
+
+        public static TSource MinBy<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector)
+        {
+            return source.MinBy(keySelector, null);
+        }
+
+        public static TSource MinBy<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IComparer<TKey> comparer)
+        {
+            if (source == null) throw Error.ArgumentNull("source");
+            if (keySelector == null) throw Error.ArgumentNull("keySelector");
+            if (comparer == null) comparer = Comparer<TKey>.Default;
+            TSource value = default(TSource);
+            if (value == null)
+            {
+                using (IEnumerator<TSource> e = source.GetEnumerator())
+                {
+                    do
+                    {
+                        if (!e.MoveNext()) return value;
+                        value = e.Current;
+                    }
+                    while (value == null);
+                    TKey key = keySelector(value);
+                    while (e.MoveNext())
+                    {
+                        TSource x = e.Current;
+                        if(x != null)
+                        {
+                            TKey xKey = keySelector(x);
+                            if (comparer.Compare(xKey, key) < 0)
+                            {
+                                value = x;
+                                key = xKey;
+                            }
+                        }
+                    }
+                }
+            }
+            else
+            {
+                using(IEnumerator<TSource> e = source.GetEnumerator())
+                {
+                    if (!e.MoveNext()) throw Error.NoElements();
+                    TKey key = keySelector(value = e.Current);
+                    while (e.MoveNext())
+                    {
+                        TSource x = e.Current;
+                        TKey xKey = keySelector(x);
+                        if (comparer.Compare(xKey, key) < 0)
+                        {
+                            value = x;
+                            key = xKey;
+                        }
+                    }
+                }
+            }
+
+            return value;
+        }
+
+        public static IEnumerable<TSource> DistinctBy<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector)
+        {
+            return source.DistinctBy(keySelector, null);
```

Comment:
I have not had any coffee yet but does null get a value later? I prefer passing EqualityComparer<T>.Default same as above for readability and then comparer can be checked for null in the overload.